### PR TITLE
Restore debugger state

### PR DESCRIPTION
**Pull request by JohanMabille**
This also includes a first pass on `IDebugger`, `IDebugger.ISession`, `DebugService` and `DebugSession` according to #139

### CODE REVIEW

File: src/session.ts [hunk 1]
```diff
@@ -24,29 +25,39 @@ export class DebugSession implements IDebugger.ISession {
...
+   * @param client - The new API client session.
+   */
   set client(client: IClientSession | null) {
     if (this._client === client) {
       return;
     }
 
     if (this._client) {
-      Signal.clearData(this._client);
+      this._client.iopubMessage.disconnect(this._handleEvent, this);
```
Comment (jtpio) [inline src/session.ts]: And `Signal.clearData(this._client.iopubMessage)` would not be suitable because it would clear too much?

Comment (JohanMabille) [reply src/session.ts]: Yes, we don't want to remove other signals, only the one of the debugger session.

Comment (JohanMabille) [reply src/session.ts]: Indeed, we don't want to remove any other slot, only the one of this session.

Comment (jtpio) [reply src/session.ts]: Good, thanks for the confirmation :+1:

**Comment by jtpio** [comment]
Thanks!

Merging so we can iterate on the restore flow.

**Pull request merged**